OSDOCS 11250: Added securing routes with externally managed certificate section in cert-manager (#78500)
openshift/openshift-docs: modules/cert-manager-configuring-routes.adoc

* You have a `Service` that you want to expose.

.Procedure

. Create a `route` for your service using edge TLS termination and a custom `hostname`.
+
[source, terminal]
----
$ oc create route edge <route_name> \ # <1>
  --service=<service_name> \ # <2>
  --hostname=<host_name> \ # <3>
  --namespace=<namespace> # <4>
----
<1> Specify your route's name.
<2> Specify the service you want to expose.
<3> Specify the hostname of your route.
<4> Specify the namespace where your route is located.

. Create an `Issuer` to configure the http01 solver. For other ACME issuer types, see "Configuring ACME an issuer" in the "Additional Resources" section.
+
.Example `Issuer.yaml` file
+
[source, yaml]
----
$ oc create -f - << EOF
apiVersion: cert-manager.io/v1
kind: Issuer
metadata:
  name: letsencrypt-acme
  namespace: <namespace> # <1>
spec:
  acme:
    server: https://acme-v02.api.letsencrypt.org/directory
    privateKeySecretRef:
      name: letsencrypt-acme-account-key
    solvers:
      - http01:
          ingress:
            ingressClassName: openshift-default
EOF
----
<1> Specify the namespace where the Issuer is located. It should be the same as your route's namespace.

> . Create a `Certificate` for the route.

Reviewer: ```suggestion
. Create a `Certificate` for the route. The `secretName` specifies the TLS secret that is going to be issued and managed by cert-manager and will also be referenced in your route in the following steps.
```

Same as above.:)

<2> Specify the certificate's common name using the `hostname` of the route.
<3> Add the `hostname` of your route to the certificate's DNS names.
<4> Specify the name of the secret that contains the certificate.

. Create a `Role` to provide the router service account permissions to read the referenced secret using the following command:
+
[source, terminal]
----
$ oc create role secret-reader \
  --verb=get,list,watch \
  --resource=secrets \
  --resource-name=<secret_name> \ # <1>
  --namespace=<namespace> # <2>
----
<1> Specify the name of the secret that you want to grant access to. It should be consistent with your `secretName` specified in the Certificate.
<2> Specify the namespace where both your secret and route reside.

. Create a `RoleBinding` to bind the router service account with the newly created `Role` using the following command:
+
[source, terminal]
----
$ oc create rolebinding secret-reader-binding \
  --role=secret-reader \
  --serviceaccount=openshift-ingress:router \
  --namespace=<namespace> # <1>
----
<1> Specify the namespace where both your secret and route reside.

> . Update your route's `.spec.tls.externalCertificate` field to reference the previously created secret and use the certificate issued by {cert-manager-operator} using the following command:

Reviewer: ```suggestion
. Update your route's `.spec.tls.externalCertificate` field to reference the previously created secret and use the certificate issued by cert-manager using the following command:
```
Here using `cert-manager` is ok.

+

> [source, terminal]
> ----
> $ oc patch route <route_name> \
>   -n <namespace> \
>   --type=merge \
>   -p '{"spec":{"tls":{"externalCertificate":{"name":"<secret_name>"}}}}'
> ----

Reviewer: ```suggestion
[source, terminal]
----
$ oc patch route <route_name> \
  -n <namespace> \ # <1>
  --type=merge \
  -p '{"spec":{"tls":{"externalCertificate":{"name":"<secret_name>"}}}}' # <2>
----
<1> Specify the namespace where both your secret and route reside.
<2> Specify the name of the secret that contains the certificate.
```

.Verification

* Verify that the certificate is created and ready to use by running the following command:
+ 
[source, terminal]
----
$ oc get certificate -n <namespace>
$ oc get secret -n <namespace>
----

* Verify that the router is using the referenced external certificate by running the following command. The command should return with the status code `200 OK`.
+ 
[source, terminal]
----
$ curl -IsS https://<hostname>
----

* Verify the server certificate's `subject`, `subjectAltName` and `issuer` are all as expected from the curl verbose outputs by running the following command:
+ 
[source, terminal]
----
$ curl -v https://<hostname>
----
+

> Now the route is successfully secured by the certificate from the referenced secret issued by {cert-manager-operator}. Cert-manager will automatically manage the certificate's lifecycle.

Reviewer: ```suggestion
Now the route is successfully secured by the certificate from the referenced secret issued by cert-manager. cert-manager will automatically manage the certificate's lifecycle.
```

It's ok to use `cert-manager` here. (no need for capital letters)

// Module included in the following assemblies:
//
// * security/cert_manager_operator/cert-manager-creating-certificate.adoc

:_mod-docs-content-type: PROCEDURE
[id="cert-manager-configuring-routes_{context}"]
= Configuring certificates to secure routes in your cluster

The following steps demonstrate the process of utilizing the {cert-manager-operator} with the _Let's Encrypt_ ACME HTTP-01 challenge type to secure the route resources in your OpenShift Cluster.

.Prerequisites

* You have installed the {cert-manager-operator} 1.14.0 or later.
* You have enabled the `RouteExternalCertificate` feature gate.
* You have the `create` and `update` permissions on the `routes/custom-host` sub-resource.
* You have a `Service` that you want to expose.

.Procedure

> . Create a `route` for your service using edge TLS termination and a custom `hostname`. The `hostname` will be used in your Certificate's definition file in the following steps.

Reviewer 2: Nitpick:
```suggestion
. Create a `route` for your `service` using edge TLS termination and a custom `hostname`. The `hostname` will be used while creating a `Certificate` in the following steps.
```

+
[source, terminal]
----
$ oc create route edge <route_name> \ # <1>
  --service=<service_name> \ # <2>
  --hostname=<host_name> \ # <3>
  --namespace=<namespace> # <4>
----
<1> Specify your route's name.
<2> Specify the service you want to expose.
<3> Specify the hostname of your route.
<4> Specify the namespace where your route is located.

. Create an `Issuer` to configure the http01 solver. For other ACME issuer types, see "Configuring ACME an issuer" in the "Additional Resources" section.
+
.Example `Issuer.yaml` file
+
[source, yaml]
----
$ oc create -f - << EOF
apiVersion: cert-manager.io/v1
kind: Issuer
metadata:
  name: letsencrypt-acme
  namespace: <namespace> # <1>
spec:
  acme:
    server: https://acme-v02.api.letsencrypt.org/directory
    privateKeySecretRef:
      name: letsencrypt-acme-account-key
    solvers:
      - http01:
          ingress:
            ingressClassName: openshift-default
EOF
----

> <1> Specify the namespace where the Issuer is located. It should be the same as your route's namespace.

Reviewer 2: nit
```suggestion
<1> Specify the namespace where the `Issuer` is located. It should be the same as your route's namespace.
```

+
[source, yaml]
----
$ oc create -f - << EOF
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: example-route-cert
  namespace: <namespace> # <1>
spec:
  commonName: <hostname> # <2>
  dnsNames:
    - <hostname> # <3>
  usages:
    - server auth
  issuerRef:
    kind: Issuer
    name: letsencrypt-acme
  secretName: <secret_name> # <4>
EOF
----
<1> Specify the `namespace` where the `Certificate` is located. It should be the same as your route's namespace.
<2> Specify the certificate's common name using the `hostname` of the route.
<3> Add the `hostname` of your route to the certificate's DNS names.
<4> Specify the name of the secret that contains the certificate.

. Create a `Role` to provide the router service account permissions to read the referenced secret using the following command:
+
[source, terminal]
----
$ oc create role secret-reader \
  --verb=get,list,watch \
  --resource=secrets \
  --resource-name=<secret_name> \ # <1>
  --namespace=<namespace> # <2>
----

> <1> Specify the name of the secret that you want to grant access to. It should be consistent with your `secretName` specified in the Certificate.

Reviewer 2: ```suggestion
<1> Specify the name of the secret that you want to grant access to. It should be consistent with your `secretName` specified in the `Certificate`.
```

<2> Specify the namespace where both your secret and route reside.

. Create a `RoleBinding` to bind the router service account with the newly created `Role` using the following command:
+
[source, terminal]
----
$ oc create rolebinding secret-reader-binding \
  --role=secret-reader \
  --serviceaccount=openshift-ingress:router \
  --namespace=<namespace> # <1>
----
<1> Specify the namespace where both your secret and route reside.

. Update your route's `.spec.tls.externalCertificate` field to reference the previously created secret and use the certificate issued by cert-manager using the following command:
+
[source, terminal]
----

> $ oc patch route <route_name> \

Reviewer 2: `<route_name>` also needs to be numbered. This will change next sequence numbers as well. 
```suggestion
$ oc patch route <route_name> \ <1>
```

  -n <namespace> \ # <1>
  --type=merge \
  -p '{"spec":{"tls":{"externalCertificate":{"name":"<secret_name>"}}}}' # <2>
----
<1> Specify the namespace where both your secret and route reside.
<2> Specify the name of the secret that contains the certificate.

.Verification

* Verify that the certificate is created and ready to use by running the following command:
+ 
[source, terminal]
----

> $ oc get certificate -n <namespace>
> $ oc get secret -n <namespace>

Reviewer 2: ```suggestion
$ oc get certificate -n <namespace> <1>
$ oc get secret -n <namespace> <1>
```

----

* Verify that the router is using the referenced external certificate by running the following command. The command should return with the status code `200 OK`.
+ 
[source, terminal]
----

> $ curl -IsS https://<hostname>

Reviewer 2: ```suggestion
$ curl -IsS https://<hostname> <1>
```

----

* Verify the server certificate's `subject`, `subjectAltName` and `issuer` are all as expected from the curl verbose outputs by running the following command:
+ 
[source, terminal]
----

> $ curl -v https://<hostname>

Reviewer 2: ```suggestion
$ curl -v https://<hostname> <1>
```

// Module included in the following assemblies:
//
// * security/cert_manager_operator/cert-manager-creating-certificate.adoc

:_mod-docs-content-type: PROCEDURE
[id="cert-manager-configuring-routes_{context}"]
= Configuring certificates to secure routes in your cluster

The following steps demonstrate the process of utilizing the {cert-manager-operator} with the _Let's Encrypt_ ACME HTTP-01 challenge type to secure the route resources in your OpenShift Cluster.

.Prerequisites

* You have installed the {cert-manager-operator} 1.14.0 or later.
* You have enabled the `RouteExternalCertificate` feature gate.
* You have the `create` and `update` permissions on the `routes/custom-host` sub-resource.

> * You have a `Service` that you want to expose.

Reviewer 3: ```suggestion
* You have a `Service` resource that you want to expose.
```
or
```suggestion
* You have a service that you want to expose.
```

.Procedure

. Create a `route` for your `service` using edge TLS termination and a custom `hostname`. The `hostname` will be used while creating a `Certificate` in the following steps.
+
[source, terminal]
----
$ oc create route edge <route_name> \ # <1>
  --service=<service_name> \ # <2>

>   --hostname=<host_name> \ # <3>

Reviewer 3: ```suggestion
  --hostname=<hostname> \ # <3>
```
per IBM SG and to match usage with the rest of the PR

  --namespace=<namespace> # <4>
----
<1> Specify your route's name.
<2> Specify the service you want to expose.
<3> Specify the hostname of your route.
<4> Specify the namespace where your route is located.

> . Create an `Issuer` to configure the http01 solver. For other ACME issuer types, see "Configuring ACME an issuer" in the "Additional Resources" section.

Reviewer 3: Or maybe this should be:

```
HTTP-01 solver
```
?

+
.Example `Issuer.yaml` file
+
[source, yaml]
----
$ oc create -f - << EOF
apiVersion: cert-manager.io/v1
kind: Issuer
metadata:
  name: letsencrypt-acme
  namespace: <namespace> # <1>
spec:
  acme:
    server: https://acme-v02.api.letsencrypt.org/directory
    privateKeySecretRef:
      name: letsencrypt-acme-account-key
    solvers:
      - http01:
          ingress:
            ingressClassName: openshift-default
EOF
----
<1> Specify the namespace where the `Issuer` is located. It should be the same as your route's namespace.

. Create a `Certificate` for the route. The `secretName` specifies the TLS secret that is going to be issued and managed by cert-manager and will also be referenced in your route in the following steps.
+
[source, yaml]
----
$ oc create -f - << EOF
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: example-route-cert
  namespace: <namespace> # <1>
spec:
  commonName: <hostname> # <2>
  dnsNames:
    - <hostname> # <3>
  usages:
    - server auth
  issuerRef:
    kind: Issuer
    name: letsencrypt-acme
  secretName: <secret_name> # <4>
EOF
----

> <1> Specify the `namespace` where the `Certificate` is located. It should be the same as your route's namespace.

Reviewer 3: ```suggestion
<1> Specify the `namespace` where the `Certificate` resource is located. It should be the same as your route's namespace.
```

> <2> Specify the certificate's common name using the `hostname` of the route.
> <3> Add the `hostname` of your route to the certificate's DNS names.

Reviewer 3: ```suggestion
<2> Specify the certificate's common name using the hostname of the route.
<3> Add the hostname of your route to the certificate's DNS names.
```

<4> Specify the name of the secret that contains the certificate.

> . Create a `Role` to provide the router service account permissions to read the referenced secret using the following command:

Reviewer 3: ```suggestion
. Create a `Role` to provide the router service account permissions to read the referenced secret by using the following command:
```

+
[source, terminal]
----
$ oc create role secret-reader \
  --verb=get,list,watch \
  --resource=secrets \
  --resource-name=<secret_name> \ # <1>
  --namespace=<namespace> # <2>
----

> <1> Specify the name of the secret that you want to grant access to. It should be consistent with your `secretName` specified in the `Certificate`.

Reviewer 3: ```suggestion
<1> Specify the name of the secret that you want to grant access to. It should be consistent with your `secretName` specified in the `Certificate` resource.
```

> <2> Specify the namespace where both your secret and route reside.

Reviewer 3: IBM SG says to avoid "reside" and use something simpler.
```suggestion
<2> Specify the namespace where both your secret and route are located.
```
There are 3 other instances of "reside" in the PR that can be updated to match as well.

> . Create a `RoleBinding` to bind the router service account with the newly created `Role` using the following command:

Reviewer 3: ```suggestion
. Create a `RoleBinding` resource to bind the router service account with the newly created `Role` resource by using the following command:
```

+
[source, terminal]
----
$ oc create rolebinding secret-reader-binding \
  --role=secret-reader \
  --serviceaccount=openshift-ingress:router \
  --namespace=<namespace> # <1>
----
<1> Specify the namespace where both your secret and route reside.

> . Update your route's `.spec.tls.externalCertificate` field to reference the previously created secret and use the certificate issued by cert-manager using the following command:

Reviewer 3: ```suggestion
. Update your route's `.spec.tls.externalCertificate` field to reference the previously created secret and use the certificate issued by cert-manager by using the following command:
```

+
[source, terminal]
----
$ oc patch route <route_name> \ # <1>
  -n <namespace> \ # <2>
  --type=merge \
  -p '{"spec":{"tls":{"externalCertificate":{"name":"<secret_name>"}}}}' # <3>
----
<1> Specify the route name.
<2> Specify the namespace where both your secret and route reside.
<3> Specify the name of the secret that contains the certificate.

.Verification

* Verify that the certificate is created and ready to use by running the following command:
+ 
[source, terminal]
----
$ oc get certificate -n <namespace> <1>
$ oc get secret -n <namespace> <1>
----
<1> Specify the namespace where both your secret and route reside.

* Verify that the router is using the referenced external certificate by running the following command. The command should return with the status code `200 OK`.
+ 
[source, terminal]
----
$ curl -IsS https://<hostname> <1>
----

> <1> Specify the `hostname` of your route.

Reviewer 3: ```suggestion
<1> Specify the hostname of your route.
```

* Verify the server certificate's `subject`, `subjectAltName` and `issuer` are all as expected from the curl verbose outputs by running the following command:
+ 
[source, terminal]
----
$ curl -v https://<hostname> <1>
----
<1> Specify the `hostname` of your route.

> +
> Now the route is successfully secured by the certificate from the referenced secret issued by cert-manager. cert-manager will automatically manage the certificate's lifecycle.

Reviewer 3: Should this be attached to the previous (third) bullet? Or should it be decoupled and apply to the overall "Verification" list.

Also suggest moving the "now" in the sentence a little:
```suggestion

The route is now successfully secured by the certificate from the referenced secret issued by cert-manager. cert-manager will automatically manage the certificate's lifecycle.
```
